### ISSUE TITLE
[bug] merge refs not found in remote when capturing commit coordinates

### ISSUE THREAD

**Issue by DukeOfKirkcaldy** [conan-io/conan#10864]
I am building packages in a pull request verification workflow to ensure that the changes integrate into all consumers.  When the sources are initially checked out, the standard behaviour of Github workflows (checkout action) is to check out the 'merge ref' (i.e. the commit which would be created if the pull request is merged).  

When capturing the commit coordinates using methods in `conan.tools.scm.Git`, conan cannot find these merge refs and falls back to return the local root path instead.  I would still like the merge ref hash to be stored so I can simply promote the same pacakges to another repo when the pull request is merged.

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10
  * Compiler+version:
  * Conan version: 1.46.0
  * Python version: 3.7.9

### Logs (Executed commands with output) (Include/Attach if Applicable)
Checkout action:
```
Checking out the ref
  /usr/bin/git checkout --progress --force refs/remotes/pull/1/merge
  Note: checking out 'refs/remotes/pull/1/merge'.
  
  You are in 'detached HEAD' state. You can look around, make experimental
  changes and commit them, and you can discard any commits you make in this
  state without impacting any branches by performing another checkout.
  
  If you want to create a new branch to retain commits you create, you may
  do so (now or later) by using -b with the checkout command again. Example:
  
    git checkout -b <new-branch-name>
  
  HEAD is now at 3d653a0 Merge 7035bcc5162cda71473312dc3670848131af40a3 into cf607d849b3d129d88f1dbfb75fc5628f7909cae
```
Conan output:
```
liba/0.1: WARN: Current commit 3d653a0222f675933f8110ecd3d19c658f0febab doesn't exist in remote origin
This revision will not be buildable in other computer
```

**Comment by memsharded** [comment]
Hi @GordonJess 

But this is expected, if the commit doesn't exist in the server, it cannot be cloned from the server, it needs to be cloned from the place where the commit exists, so this is the rationale of the current proposed approach.

It seems that you want a special, different behavior depending on a future event (pushing that commit to the remote) that has happened yet, and then Conan doesn't have the necessary information to work. So probably the approach should be something like:

- In the ``export()`` method, capture both the local URL and the remote URL. Capture the commit too.
- In the ``source()`` method, try both checkouts in order, first the local URL, and if the commit is not there, then the remote URL

Do you think that would work? It should be possible to implement it with the same proposed approach (``export()`` to store data in conandata.yml, ``source()`` to retrieve the source)

**Comment by DukeOfKirkcaldy** [comment]
Hi @memsharded.

Thanks for the suggestion but I think there is a small understanding.  The commit **does** exist in the server, but it is not possible to check it out directly because it is not found through the default refspec.  The logs of the Github Action show that the commit is fetched like so:

`git fetch origin +3d653a0222f675933f8110ecd3d19c658f0febab:refs/remotes/pull/1/merge`

I could probably add a similar fetch in the `source()` method to ensure it can be checked out.

But the `export()` part still fails - not because the commit is not found, but [because the ref does not include origin](https://github.com/conan-io/conan/blob/f69c0269f9b7b5842f12de2ae444212b9a80a896/conan/tools/scm/git.py#L42).

**Comment by DukeOfKirkcaldy** [comment]
Just to confirm, I can get the `source()` method to work by fetching an additional refspec:

conandata.yml:
```
sources:
  commit: 3d653a0222f675933f8110ecd3d19c658f0febab
  url: git@github.com:MY-ORG/conan-demo-liba
```
conanfile.py:
```python
def source(self):
    git = Git(self)
    # recover the saved url and commit from conandata.yml and use them to get sources
    sources = self.conan_data["sources"]
    git.clone(url=sources["url"], target=".")
    git._run('fetch origin +refs/pull/*:refs/remotes/pull/*')
    git.checkout(commit=sources["commit"])
```
---

Would be neat if this new Git helper's `clone()` method could also accept `args` so I could add it as a config:
```python
pr_refspec = "-c remote.origin.fetch +refs/pull/*:refs/remotes/pull/*" if EnvVars.get("GITHUB_ACTIONS") else ""
git.clone(url=sources["url"], target=".", args=pr_refspec)
```

**Comment by memsharded** [comment]
Wouldn't it be cleaner a `git.fetch()`` that can be used like?:
```python
git.clone(...)
if <github-action>: 
    git.fetch("origin", "+refs/pull/*:refs/remotes/pull/*")
```

Regarding the detection, what would be the command to be used?
Right now as you pointed, we are doing something like:
```python
 def commit_in_remote(self, commit, remote="origin"):
      branches = self._run("branch -r --contains {}".format(commit))
      return "{}/".format(remote) in branches
```
What would be a similar command that would allow to test that such commit exists in the remote?

**Comment by DukeOfKirkcaldy** [comment]
Yeah that would be better actually.  I thought the checkout ref could be added to the clone command (using -b <commit>), and in that case the refspec would need to be added as config.

Anyway, I think as long as branches is not empty then the commit is found, right?

**Comment by DukeOfKirkcaldy** [comment]
In case anyone else has this issue, I was able to workaround it in my CI workflow by doing an addition refspec fetch after checkout, to map the merge commit to a ref containing `origin`:

```
git fetch --depth=1 origin +${{ github.sha }}:refs/remotes/origin/${{ github.ref_name }}
```

It's not ideal and doesn't work for local development, but can get by with it for now!
